components/cloud/azure/src/token_credentials_with_cache/certificate_credentials.rs

```diff
@@ -0,0 +1,283 @@
+/// `ClientCertificateCredential` if `ClientCertificateCredential` is nightly
+/// released.
+pub struct ClientCertificateCredentialWithCache {
+    http_client: Arc<dyn HttpClient>,
+
+    tenant_id: String,
+    client_id: String,
+    /// Certificate in PKCS12 format, encoded in base64
+    certificate: String,
+    /// Certificate Pass, default with ""
+    certificate_pass: String,
+    options: CertificateCredentialOptions,
+
+    /// Cached token
+    cached_token: TokenCache,
+}
+
+impl ClientCertificateCredentialWithCache {
+    /// Create a new ClientCertificateCredentialWithCache
+    pub fn new(
+        tenant_id: String,
+        client_id: String,
+        certificate: String,
+        certificate_pass: String,
+    ) -> Self {
+        Self {
+            http_client: new_http_client(),
+            tenant_id,
+            client_id,
+            certificate,
+            certificate_pass,
+            options: CertificateCredentialOptions::default(),
+            cached_token: TokenCache::default(),
+        }
+    }
+
+    /// Build a new ClientCertificateCredentialWithCache according to
+    /// a given certificate.
+    pub fn build(
+        tenant_id: String,
+        client_id: String,
+        certificate_path: String,
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        let bytes = std::fs::read(certificate_path)?;
+        Ok(ClientCertificateCredentialWithCache::new(
+            tenant_id,
+            client_id,
+            base64::encode(bytes),
+            "".into(),
+        ))
+    }
+
+    fn options(&self) -> &CertificateCredentialOptions {
+        &self.options
+    }
+
+    fn sign(jwt: &str, pkey: &PKey<Private>) -> Result<Vec<u8>, ErrorStack> {
+        let mut signer = Signer::new(MessageDigest::sha256(), pkey)?;
+        signer.update(jwt.as_bytes())?;
+        signer.sign_to_vec()
+    }
+
+    fn get_thumbprint(cert: &X509) -> Result<DigestBytes, ErrorStack> {
+        let der = cert.to_der()?;
+        let digest = hash(MessageDigest::sha1(), &der)?;
+        Ok(digest)
+    }
+
+    fn as_jwt_part(part: &[u8]) -> String {
+        base64::encode_url_safe(part)
+    }
+}
+
+impl TokenCredentialWithCache for ClientCertificateCredentialWithCache {
+    fn need_update_cached_token(&self) -> bool {
+        self.cached_token.need_update_token()
+    }
+
+    fn update_cached_token(&self, token: TokenResponse) {
+        self.cached_token.update_token(token);
+    }
+
+    fn get_cached_token(&self) -> TokenResponse {
+        self.cached_token.get_token()
+    }
+}
+
+fn get_encoded_cert(cert: &X509) -> azure_core::Result<String> {
+    Ok(format!(
+        "\"{}\"",
+        base64::encode(cert.to_pem().map_err(openssl_error)?)
+    ))
+}
+
+fn openssl_error(err: ErrorStack) -> azure_core::error::Error {
+    Error::new(ErrorKind::Credential, err)
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
```

Comment:
Why do we care about "wasm32"?

Reply:
It's used to meet the requirement of the trait `TokenCredential`. As this feature has been supported in the latest version of `Azure-sdk-for-rust`, this part on caching `TokenCredential` can be skipped.

Reply:
Could you add some comments?

Reply:
Done for it ,thx.